src/Features/Core/Portable/CustomMessageHandler/CustomMessageHandlerService.cs

```diff
@@ -0,0 +1,350 @@
+        string assemblyFolderPath,
+        CancellationToken cancellationToken)
+    {
+#if DEBUG
+        System.Diagnostics.Debugger.Launch();
+#endif
+
+        try
+        {
+            CustomMessageHandlerExtension extension = default;
+            lock (_lockObject)
+            {
+                if (_extensions.TryGetValue(assemblyFolderPath, out extension))
+                {
+                    _extensions.Remove(assemblyFolderPath);
+
+                    foreach (var assembly in extension.Assemblies)
+                    {
+                        if (assembly.Value.HasValue)
+                        {
+                            foreach (var handler in assembly.Value.Value.Handlers)
+                            {
+                                _handlers.Remove(handler);
+                            }
+
+                            foreach (var documentHandler in assembly.Value.Value.DocumentHandlers)
+                            {
+                                _documentHandlers.Remove(documentHandler);
+                            }
+                        }
+                    }
+                }
+            }
+
+            extension.AssemblyLoadContext?.Unload();
+        }
+        catch (Exception e) when (LogAndPropagate(e))
+        {
+            // unreachable
+        }
+
+        bool LogAndPropagate(Exception e)
+        {
+            Logger.Log(
+                FunctionId.CustomMessageHandlerService_UnloadCustomMessageHandlerAsync,
+                $"Error unloading {assemblyFolderPath}: {e}",
+                LogLevel.Error);
+            return false;
+        }
+
+        return ValueTask.CompletedTask;
+    }
+
+    public ValueTask ResetAsync(CancellationToken cancellationToken)
+    {
+        Clear();
+        return ValueTask.CompletedTask;
+    }
+
+    public void Dispose()
+        => Clear();
+
+    private void Clear()
+    {
+        List<CustomMessageHandlerExtension> extensions;
+        lock (_lockObject)
+        {
+            extensions = _extensions.Values.ToList();
+            _extensions.Clear();
+            _handlers.Clear();
+            _documentHandlers.Clear();
+        }
+
+        foreach (var extension in extensions)
+        {
+            extension.AssemblyLoadContext.Unload();
+        }
+    }
+
+    private readonly struct CustomMessageHandlerExtension(AssemblyLoadContext assemblyLoadContext) : IEquatable<CustomMessageHandlerExtension>
+    {
+        public AssemblyLoadContext AssemblyLoadContext { get; } = assemblyLoadContext;
+
+        public Dictionary<string, CustomMessageHandlerAssembly?> Assemblies { get; } = new();
+
+        public object AssemblyLoadLockObject { get; } = new();
+
+        // Used to avoid race conditions between RegisterHandlersAsync and UnloadCustomMessageHandlersAsync
+        public bool Equals(CustomMessageHandlerExtension other)
+            => ReferenceEquals(Assemblies, other.Assemblies);
```

Comment:
trying to figure out hwo this could ever be true between different instances.   and if this is all about just having this be true for a single instance, just make this a class, without any need for IEqutable.